### ISSUE TITLE
Please make (or document) a generic API for your optimisers/solvers

### ISSUE THREAD

**Issue by evan-wehi** [deeplearning4j/deeplearning4j#4179]
As far as I can see your optimisers are only readily available to train neural nets. It would great to able to be call them for arbitrary optimisation tasks. My particular requirement is for neural style transfer (https://arxiv.org/abs/1508.06576). In this case, an output image is reconstructed by minimising the weighted loss between features of the new image and features of the  content and style images (eq 7 in the previous reference). Here is a tensorflow implementation: https://github.com/anishathalye/neural-style/blob/b7461815c8fee73199709273c76daf69982db3ec/stylize.py#L130

I would think having the optimisers generally available would be useful in many other situations as well.

**Comment by evan-wehi** [comment]
Here are a couple more use cases that are relevant to me:

- https://arxiv.org/abs/1705.01088
- https://arxiv.org/abs/1412.0035

They all variations on the theme of using a CNN (or other model) for visual feature extraction then optimising to (re)construct an image with those features.

**Comment by agibsonccc** [comment]
Closing — this issue is over 7 years old with no recent activity. The codebase has undergone a major rewrite. If this is still relevant, please open a new issue with current reproduction steps.
